assets/js/amp-block-validation.js

```diff
@@ -0,0 +1,166 @@
+/**
+ * Validates blocks for AMP compatibility.
+ *
+ * This uses the REST API response from saving a page to find validation errors.
+ * If one exists for a block, it display it inline with a Notice component.
+ */
+
+/* exported ampBlockValidation */
+/* global wp, _ */
+var ampBlockValidation = ( function() {
+	'use strict';
+
+	var module = {
+
+		/**
+		 * Holds data.
+		 */
+		data: {
+			i18n: {
+				invalidAmpContentNotice: ''
+			}
+		},
+
+		/**
+		 * Boot module.
+		 *
+		 * @param {Object} data - Module data.
+		 * @return {void}
+		 */
+		boot: function boot( data ) {
+			var lastValidationErrors;
+			module.data = data;
+
+			wp.hooks.addFilter(
+				'blocks.BlockEdit',
+				'amp/add-notice',
+				module.conditionallyAddNotice
+			);
+
+			module.store = wp.data.registerStore( 'amp/blockValidation', {
+				reducer: function( _state, action ) {
+					var state = _state || {
+						blocksValidationErrors: {}
+					};
+
+					switch ( action.type ) {
+						case 'UPDATE_BLOCKS_VALIDATION_ERRORS':
+							return _.extend( {}, state, {
+								blocksValidationErrors: action.blocksValidationErrors
+							} );
+						default:
+							return state;
+					}
+				},
+				actions: {
+					updateBlocksValidationErrors: function( blocksValidationErrors ) {
+						return {
+							type: 'UPDATE_BLOCKS_VALIDATION_ERRORS',
+							blocksValidationErrors: blocksValidationErrors
+						};
+					}
+				},
+				selectors: {
+					getBlockValidationErrors: function( state, uid ) {
+						return state.blocksValidationErrors[ uid ] || [];
+					}
+				}
+			} );
+
+			wp.data.subscribe( function() {
+				var thisValidationErrors = wp.data.select( 'core/editor' ).getCurrentPost()[ module.data.restValidationErrorsField ];
+				if ( thisValidationErrors && ! _.isEqual( lastValidationErrors, thisValidationErrors ) ) {
+					lastValidationErrors = thisValidationErrors;
+					module.updateBlocksValidationErrors( thisValidationErrors );
+				}
+
+				// @todo wp.data.dispatch( 'core/editor' ).createNotice()?
+			} );
+		},
+
+		/**
+		 * Update blocks' validation errors in the store.
+		 *
+		 * @param {Object[]} validationErrors - Validation errors.
+		 * @return {void}
+		 */
+		updateBlocksValidationErrors: function updateBlocksValidationErrors( validationErrors ) {
+			var blocksValidationErrors, blockOrder, currentPost, selectors;
+			selectors = wp.data.select( 'core/editor' );
+			blockOrder = selectors.getBlockOrder();
```

Comment:
There is a race condition here where:

1. I add an AMP invalidity to the first block.
2. I hit Save.
3. The server is slow to respond, and while waiting to respond I move a block.
4. When the server responds, the block indices in the validation errors are no longer aligned with the order of blocks in the editor and so validation errors are mismatched.

To fix this problem, we need to capture the `blockOrder` at the moment that saving is initiated instead of using the state at the time that saving is finished.